File: topN.py
#!/usr/bin/env python

import argparse
import os
import time

from pprint import pprint as pp

def check(final, number):

    # Optimization to avoid scan the whole list if the number
    # wont make the cut
    if number < final[-1]:
        return

    # Original algorithm
    for i in final:
        if number > i:
            pos = final.index(i)
            final.insert(pos, number)
            final.pop()
            break


def topN(filename, top_numbers):
    final = [x for x in range(int(top_numbers)-1, -1, -1)]
    with open(filename, 'r') as f:
        for number in f:
            if number != '':
                check(final, int(number.strip('\n')))
    pp(final)

def main(filename, top_numbers):
    start = time.time()
    topN(filename, top_numbers)
    end = time.time()
    stat = os.stat(filename)
    print('To sort the %s highest numbers from  %s with size %s MB it took %s' % (top_numbers, filename, (stat.st_size/1048576), end-start))


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='topN: find the top N largest number from a file')
    parser.add_argument('--filename', help='file to search for top numbers', required=True)
    parser.add_argument('--top-numbers', help='number of top numbers to be retrieved', required=True)
    args = parser.parse_args()

    main(args.filename, args.top_numbers)

File: topN.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"log"
	"os"
	"strconv"
	"time"
)

var top_numbersFlag int
var filenameFlag string

func init() {
	flag.IntVar(&top_numbersFlag, "top-numbers", 0, "topN wanted to retrieve")
	flag.StringVar(&filenameFlag, "filename", "numbers.txt", "file to search for top numbers")
}

func check_number(final []int, number int) []int {
	for i, v := range final {
		if number > v {
			// https://github.com/golang/go/wiki/SliceTricks
			// pop
			_, final := final[len(final)-1], final[:len(final)-1]
			// insert
			final = append(final[:i], append([]int{number}, final[i:]...)...)
			break
		}
	}
	return final
}

func build_slice(top_numbers int) []int {
	log.Printf("Building slice of %d elements", top_numbers)
	var final []int

	for v := top_numbers; v > 1; v-- {
		final = append(final, v)
	}

	return final
}

func main() {
	flag.Parse()

	final := build_slice(top_numbersFlag)

	start := time.Now()

	file, err := os.Open(filenameFlag) // For read access.
	if err != nil {
		log.Fatal(err)
	}

	log.Printf("Starting to read file %s", filenameFlag)
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		number, err := strconv.Atoi(scanner.Text())
		if err != nil {
			log.Fatal(err)
		}
		final = check_number(final, number)
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	end := time.Now()
	fmt.Printf("The call took %v to run.\n", end.Sub(start))

	for _, v := range final {
		log.Println(v)
	}

}
